consensus/spos/scheduledProcessor.go

```diff
@@ -0,0 +1,188 @@
+	case inProgress:
+		return inProgressString
+	case processingOK:
+		return processingOKString
+	default:
+		return unexpectedString
+	}
+}
+
+// ScheduledProcessorWrapperArgs holds the arguments required to instantiate the pipelineExecution
+type ScheduledProcessorWrapperArgs struct {
+	SyncTimer                  ntp.SyncTimer
+	Processor                  process.ScheduledBlockProcessor
+	ProcessingTimeMilliSeconds uint32
+}
+
+type scheduledProcessorWrapper struct {
+	syncTimer      ntp.SyncTimer
+	processingTime time.Duration
+	processor      process.ScheduledBlockProcessor
+
+	startTime time.Time
+	status    processingStatus
+	sync.RWMutex
+
+	oneInstance sync.RWMutex
+}
+
+// NewScheduledProcessorWrapper creates a new processor for scheduled transactions
+func NewScheduledProcessorWrapper(args ScheduledProcessorWrapperArgs) (*scheduledProcessorWrapper, error) {
+	if check.IfNil(args.SyncTimer) {
+		return nil, process.ErrNilSyncTimer
+	}
+	if check.IfNil(args.Processor) {
+		return nil, process.ErrNilBlockProcessor
+	}
+	if args.ProcessingTimeMilliSeconds < 1 {
+		return nil, process.ErrInvalidProcessingTime
+	}
+
+	return &scheduledProcessorWrapper{
+		syncTimer:      args.SyncTimer,
+		processingTime: time.Duration(args.ProcessingTimeMilliSeconds) * time.Millisecond,
+		processor:      args.Processor,
+		status:         processingNotStarted,
+	}, nil
+}
+
+func (sp *scheduledProcessorWrapper) computeRemainingProcessingTime() time.Duration {
+	currTime := sp.syncTimer.CurrentTime()
+
+	sp.RLock()
+	elapsedTime := currTime.Sub(sp.startTime)
+	sp.RUnlock()
+
+	if elapsedTime < 0 {
+		return 0
+	}
+
+	return sp.processingTime - elapsedTime
+}
+
+func (sp *scheduledProcessorWrapper) waitForProcessingResult() processingStatus {
+	status := sp.getStatus()
+	for status == inProgress {
+		remainingExecutionTime := sp.computeRemainingProcessingTime()
+		select {
+		case <-time.After(remainingExecutionTime):
+			status = sp.getStatus()
+			return status
+		case <-time.After(processingCheckStep):
+			status = sp.getStatus()
+		}
+	}
+
+	return status
+}
+
+// IsProcessedOK returns true if the scheduled processing was finalized without error
+// Function is blocking until the allotted time for processing is finished
+func (sp *scheduledProcessorWrapper) IsProcessedOK() bool {
+	status := sp.waitForProcessingResult()
+
+	processedOK := status == processingOK
+	log.Debug("scheduledProcessorWrapper.IsProcessedOK", "status", status.String())
+	return processedOK
+}
+
+// StartScheduledProcessing starts the scheduled processing
+func (sp *scheduledProcessorWrapper) StartScheduledProcessing(header data.HeaderHandler, body data.BodyHandler) {
+	if !header.HasScheduledSupport() {
+		log.Debug("scheduled processing not supported")
+		sp.setStatus(processingOK)
+		return
+	}
+
+	sp.oneInstance.Lock()
+
+	sp.setStatus(inProgress)
+	log.Debug("scheduledProcessorWrapper.StartScheduledProcessing - scheduled processing has been started")
+
+	go func() {
```

Comment:
a lot of my previous questions about running in go routine were answered here. Although I thing it would be safer if this function gets a context and the go routine can be closed from the outside.

Reply:
added a forceStop for scheduled execution